### PR TITLE
Network addon session refactor

### PR DESCRIPTION
**Pull request by modmuss50**
Big thanks to @deirn for 98% of the changes in this PR, this is a follow on from #3377 to correctly handle the transition between config and play.

This PR also contains a fix for #3384 and now correctly validates the packet size in a few places where it wasnt previously. Testing still needs to be done, will likely release this in a beta.

### CODE REVIEW

File: fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/impl/networking/GlobalReceiverRegistry.java [hunk 1]
```diff
@@ -145,17 +153,29 @@ public void endSession(AbstractNetworkAddon<H> addon) {
...
 		} finally {
 			lock.unlock();
 		}
 	}
 
+	/**
+	 * In practice, trackedAddons should never contain more than one instance.
+	 */
+	private void logTrackedAddonSize() {
+		if (LOGGER.isDebugEnabled() && this.trackedAddons.size() > 1) {
```
Comment (liach) [inline fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/impl/networking/GlobalReceiverRegistry.java]: Does this `size <= 1` constraint hold true on servers, given each server player has one network handler which has one network addon?

File: fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/impl/networking/AbstractNetworkAddon.java [hunk 2]
```diff
@@ -64,13 +75,32 @@ public H getHandler(Identifier channel) {
 		}
 	}
 
+	private void assertNotReserved(Identifier channel) {
+		if (this.isReservedChannel(channel)) {
+			throw new IllegalArgumentException(String.format("Cannot register handler for reserved channel with name \"%s\"", channel));
```
Comment (liach) [inline fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/impl/networking/AbstractNetworkAddon.java]: This IAE may be also thrown if it's for unregistering; consider updating the message.

File: fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/impl/networking/GlobalReceiverRegistry.java (hunk 1 above)
Comment (deirn) [reply fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/impl/networking/GlobalReceiverRegistry.java]: Actually, yeah. I forgot to check about it 😅.

Comment (modmuss50) [reply fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/impl/networking/GlobalReceiverRegistry.java]: I think ill drop this log to trace, to only be used for testing. I dont see a nice way to make this work for servers.

**Pull request merged**